### DISCUSSION TITLE
Delete or Archive old log files through shell script

### DISCUSSION THREAD

**Discussion by joindevopscloud** [daws-76s/.github#8] [Q&A] [answered]
**Requirement is below**

- A script to delete/archive old logs in source directory, if user action is delete, delete them. If user action is archive zip them and keep it into destination directory
- User have to provide
     - Source directory
     - Action. Either archive or delete
     - If archive, he must provide destination. If delete no need of destination
     - Days. How many days old. by default 14 days
     - Memory. Optional. if user gives consider it otherwise dont consider memory
   
 **Example**
`sh old-logs.sh -s <source-dir> -a <archive|delete> -d <destination> -t <day> -m <memory-in-mb>`

**Sample**
`sh old-logs.sh -s "/tmp/shellscript-logs" -a delete  -t 7 -m 10`

**Validations**

- User has given all required input or not
- Directories are exist or not

**Comment by Satish-Kalari** [comment]
#!/bin/bash

R="\e[31m"
G="\e[32m"
Y="\e[33m"
N="\e[0m"

# Check if all arguments are provided
if [ $# -lt 4 ]; then
 echo "Usage: $0 <source_directory> <action> <destination_directory> <days> [max_size]"
 exit 1
fi

# Assign arguments to variables
source_directory=$1
action=$2
destination_directory=$3
days=$4
max_size=$5

# Check if action is either delete or archive
if [ "$action" != "delete" ] && [ "$action" != "archive" ]; 
then
 echo -e "$R Invalid action.$N Action must be either delete or archive."
 exit 1
fi

# Check if action is archive and destination directory is provided
if [ "$action" = "archive" ] && [ -z "$destination_directory" ]; 
then
 echo -e " $R Destination directory must be provided when action is archive.$N"
 exit 1
fi

# If max_size is provided, find files older than the specified number of days and larger than the maximum size
if [ -n "$max_size" ]; 
then
 files=$(find "$source_directory" -name "*.log" -mtime +$days -size +$max_size)
else
 files=$(find "$source_directory" -name "*.log" -mtime +$days)
fi

# Loop over the files
for file in $files; 
do
 # If action is delete, delete the file
 if [ "$action" = "delete" ]; then
 rm "$file"
 # If action is archive, archive the file
 elif [ "$action" = "archive" ]; then
 zip -r "$destination_directory/$(basename "$file").zip" "$file"
 rm "$file"
 fi
done

**Reply by joindevopscloud** [reply]
Here in this assignment you should not use $1, $2, etc, do the changes, this is called as optargs
`sh old-logs.sh -s <source-dir> -a <archive|delete> -d <destination> -t <day> -m <memory-in-mb>`
provide your GIT URL as well here.

**Reply by Satish-Kalari** [reply]
https://github.com/Satish-Kalari/shell-learning-.git

**Reply by joindevopscloud** [reply]
@Satish-Kalari add usage function with clear documentation inside function

**Comment by Satish-Kalari** [comment]
#!/bin/bash

R="\e[31m"
G="\e[32m"
Y="\e[33m"
N="\e[0m"

# Initialize variables
source_directory=""
action=""
destination_directory=""
days=""
max_size=""

# Parse command-line options
while getopts ":s:a:d:t:m:" opt; 
do
 case ${opt} in
   s )
     source_directory=$OPTARG
     ;;
   a )
     action=$OPTARG
     ;;
   d )
     destination_directory=$OPTARG
     ;;
   t )
     days=$OPTARG
     ;;
   m )
     max_size=$OPTARG
     ;;
   \? )
     echo -e "$R Invalid option:$N $OPTARG" 1>&2
     exit 1
     ;;
   : )
     echo -e "$R Invalid option:$N $OPTARG requires an argument" 1>&2
     exit 1
     ;;
 esac
done
shift $((OPTIND -1))

# Check if action is either delete or archive
if [ "$action" != "delete" ] && [ "$action" != "archive" ]; 
then
 echo -e "$R Invalid action.$N Action must be either delete or archive."
 exit 1
fi

# Check if action is archive and destination directory is provided
if [ "$action" = "archive" ] && [ -z "$destination_directory" ]; 
then
 echo -e " $R Destination directory must be provided when action is archive.$N"
 exit 1
fi

# If max_size is provided, find files older than the specified number of days and larger than the maximum size
if [ -n "$max_size" ]; 
then
 files=$(find "$source_directory" -name "*.log" -mtime +$days -size +$max_size)
else
 files=$(find "$source_directory" -name "*.log" -mtime +$days)
fi

# Loop over the files
for file in $files; 
do
 # If action is delete, delete the file
 if [ "$action" = "delete" ]; 
 then
 rm "$file"
 # If action is archive, archive the file
 elif [ "$action" = "archive" ]; 
 then
 zip -r "$destination_directory/$(basename "$file").zip" "$file"
 rm "$file"
 fi
done

**Comment by Sravanthich1305** [comment]
cat 01-Delete_Archive_logs.sh
#!/bin/bash

R="\e[31m"
G="\e[32m"

# Set the directory path where log files are located
log_directory="/var/log/"

# Set the threshold age for log files
threshold_days=14

# delete old log files
delete_old_logs() {
    find "$log_directory" -type f -name "*.2023*" -mtime +"$threshold_days" -exec rm {} \;
    echo "Old log files $R deleted."
}

# Function to archive old log files
archive_old_logs() {
    archive_name="log_archive_$(date +'%Y%m%d').tar.gz"
    find "$log_directory" -type f -name "*.2023*" -mtime +"$threshold_days" -exec tar -czvf "$archive_name" {} +
    echo "Old log files $G archived to $archive_name."
}

**Comment by knbob** [comment, marked as answer]
```
#!/bin/bash

**#---------------VARIABLE----------------**
SOURCE_DIR=""
DESTINATION_DIR=""
DAYS=14
MEMORY=""
ACTION=""

**#---------------- USAGE FUNCTOIN USED TO DISPLAY ------------------------------------------------**
USAGE(){
    echo "COMMAND USAGE:: $(basename $0) -s <Source Directory> -d <Destination Directory> -a <action: delete or archive> -t <No.of Days> -m<Memory size of file>"
    echo "Options: "
    echo " -s: Specify the source directory"
    echo " -d: Specify the destination directory when -a is archive"
    echo " -t: Number of days back to delete log file"
    echo " -m: Memory size of the file to delete log file"
    echo " -a: Action to be performed on log files. Options : Delete or Archive"
    echo " -h: Display Help and exit" 
}


**#----------------------OPT ARGS---------------------------------------**
while getopts ":s:d:t:m:a:h" opt;
do
    case $opt in
        s) SOURCE_DIR="$OPTARG";;
        d) DESTINATION_DIR="$OPTARG";;
        t) DAYS="$OPTARG";;
        m) MEMORY="$OPTARG";;
        a) ACTION="$OPTARG";;
        h) USAGE; exit;;
        \?) echo "Invalid Options: -"$OPTARG"" >$2; USAGE; exit;;
    esac
done

**#------------------CHECKING ACTION ARGUMENT PROVIDED OR NOT---------------**

if [ -z "$ACTION" ] 
then
    echo "ERROR:: -a option is mandatory. Either Delete or Archive"
    exit 1
fi

**#---------------------CHECKING DESTINATION DIR PROVIDED IF ARCHIVE OPTION CHOOSEN---------------**

if [ -z "$DESTINATION_DIR" ] && [ "$ACTION" == "arcive" ]
then
    echo "ERROR:: -d option is mandatory when -a is archive"
    exit 1
fi

if [ ! -d $SOURCE_DIR ] # ! denotes opposite
then
    echo "Source directory: $SOURCE_DIR does not exists."
fi

**#-------------------------DELETE/ARCHIVE ACTION----------------------**

if [ "$ACTION"=="delete" ]
then
FILES_TO_DELETE=$(find $SOURCE_DIR -type f -mtime +"$DAYS" -name "*.log")

while IFS= read -r line
do
    echo "Deleting file: $line"
    rm -rf $line
done <<< $FILES_TO_DELETE
else
FILES_TO_ARCHIVE=$(find $SOURCE_DIR -type f -mtime +"$DAYS" -name "*.log")

while IFS= read -r line
do
    echo "Archiving file: $line"
    zip -r "$DESTINATION_DIR/$(basename "$line").zip"
    rm -rf $line
done <<< $FILES_TO_ARCHIVE
fi
```

**Reply by joindevopscloud** [reply]
@knbob mention your script URL here

**Reply by joindevopscloud** [reply]
@knbob test this script and add the screenshots too if possible

**Reply by knbob** [reply]
![2023-12-23_071857](https://github.com/daws-76s/.github/assets/153567059/c0e05d92-bb24-487f-88c5-7275028e6d77)
![2023-12-23_072324](https://github.com/daws-76s/.github/assets/153567059/4747f539-f3f1-42db-9812-9582335daf60)
![2023-12-23_072431](https://github.com/daws-76s/.github/assets/153567059/802b116e-ca25-44e1-b25e-9f4480f95ee9)
![2023-12-23_072507](https://github.com/daws-76s/.github/assets/153567059/d7052c70-d439-4dcb-bfc3-278cef41e9a4)
![2023-12-23_074726](https://github.com/daws-76s/.github/assets/153567059/d87240a3-b05e-45d5-ad27-6c940f719837)

**Reply by MachaSreenath** [reply]
Title: Logical Issue in Archiving Section of Script  

The script currently zips files and deletes the originals without notifying users, which might lead to unintended data loss. Consider adding a notification or making file deletion optional to enhance clarity and user control.

Hi @knbob ,  

I noticed a potential logical issue in the script regarding the archiving functionality:  

--> In the archiving section, the script zips the files and then deletes the originals using `rm -rf`.  
-->This might lead to unintended data loss if users expect the original files to be preserved after archiving.  

  Suggestion:
--> Notify users about this behavior (archiving + deletion) explicitly.  
--> Optionally, allow flexibility by making file deletion after archiving an optional step, based on a user-provided flag.  

This is a minor update but can enhance user clarity and prevent accidental data loss.  


And another one is spelling mistake is there "arcive"  in 47th line. If user gives archive then it comes error.

Thank you

**Comment by hemanth3509** [comment]
Created the script for deleteing and archive the old logs.

Github link --> https://github.com/hemanth3509/one-repo.git

Documented the test cases -->
[Delete and Archive the old log files.docx](https://github.com/daws-76s/.github/files/13758443/Delete.and.Archive.the.old.log.files.docx)

Let me know incase any modifications needs to be done.

Regards,
Hemanth
